pandas/core/groupby/groupby.py

```diff
@@ -2984,9 +2984,17 @@ def _reindex_output(
             return output
 
         levels_list = [ping.group_index for ping in groupings]
-        index, _ = MultiIndex.from_product(
-            levels_list, names=self.grouper.names
-        ).sortlevel()
+        try:
+            index, _ = MultiIndex.from_product(
+                levels_list, names=self.grouper.names
+            ).sortlevel()
+        except ValueError as err:
+            raise ValueError(
+                "Product space too large to allocate arrays! You're trying to "
+                "group over too complex a column cross-product to fit in memory. "
```

Comment:
What does it mean to be complex? I think large may be a better description. Also, if the message were to start with "Group by cross product space", could we remove the entire second sentence?